### PR TITLE
localclient: pass input args/kwargs through salt.utils.args.parse_input

### LINKED ISSUE #99: pepper vs. salt has different `test.arg_repr` output types
When using `pepper` versus the standard `salt` command you get different results in terms of returned data types. 

This seems related to #57 but different in the details.

```bash
(envsalt)  >>> pepper -a pam --username=saltvirt --password=saltvirt 'sip-cannon' test.arg_repr trunktest2 2 '2048' image=/var/lib/libvit/images/centos66.qcow2 enable_vnc=True config='{"master": "goodboy.qa.sangoma.local"}' start=False enable_qcow=True                     
{
    "return": [
        {
            "sip-cannon": {
                "args": "('trunktest2', '2', '2048')",
                "kwargs": "{'__pub_user': 'saltvirt', '__pub_arg': ['trunktest2', '2', '2048', 'image=/var/lib/libvirt/images/centos66.qcow2', 'enable_vnc=True', 'config={\"master\": \"goodboy.qa.sangoma.local\"}', 'start=False', 'enable_qcow=True'], 'start': False, 'config
': {'master': 'goodboy.qa.sangoma.local'}, 'image': '/var/lib/libvirt/images/centos66.qcow2', '__pub_fun': 'test.arg_repr', 'enable_qcow$
: True, '__pub_jid': '20170301004836803665', '__pub_tgt': 'sip-cannon', '__pub_tgt_type': 'glob', 'enable_vnc': True, '__pub_ret': ''}"
            }
        }
    ]
}
 
(envsalt)  >>> sudo salt 'sip-cannon' test.arg_repr trunktest2 2 2048 image=/var/lib/libvirt/images/centos66.qcow2 enable_vnc=True con$
ig='{"master": "goodboy.qa.sangoma.local"}' start=False enable_qcow=True
sip-cannon:
    ----------
    args:
        ('trunktest2', 2, 2048)
    kwargs:
        {'__pub_user': 'sudo_tyler', '__pub_arg': ['trunktest2', 2, 2048, {'enable_qcow': True, 'image': '/var/lib/libvirt/images/centos66.qcow2', 'start': False, 'config': {'master': 'goodboy.qa.sangoma.local'}, 'enable_vnc': True}], 'enable_qcow': True, 'config': {'master': 'goodboy.qa.sangoma.local'}, 'image': '/var/lib/libvirt/images/centos66.qcow2', '__pub_fun': 'test.arg_repr', 'start': False, '__pub_id': '20170301004859437566', '__pub_tgt': 'sip-cannon', '__pub_tgt_type': 'glob', 'enable_vnc': True, '__pub_ret': ''}
```

Notably the `arg` array contains different types (in this case most obviously `int` versus `str`),
- `pepper`:
```python
('trunktest2', '2', '2048')
```

- `salt`: 
```python
('trunktest2', 2, 2048)
```

In my case this is turn causes a parsing problem downstream in the `salt.modules.virt.init()` function:
```
00:34:44 [ 2382] [DEBUG   ] Generating VM XML                                                                                            
libvirt: Domain Config error : XML error: Invalid value '20482048204820482048204820482048204820482048204820482048204820482048204820482048
20482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482
04820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820
48204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204
82048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048
20482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482
04820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820
482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204820482048204                                
```

### PR DESCRIPTION
**Pull request by mattp-**
### What does this PR do?
this functionality brings localclient in line with what happens on the
cli, as well as what runnerclient does. this ensures that any foo=bar
kwargs passed in as args are handled accordingly. right now localclient
relies on SaltCMDOptionParser to do this handling for us, but that
leaves a hole when salt-api interactions are brought in.

### What issues does this PR fix or reference?
should fix https://github.com/saltstack/pepper/issues/99

### Previous Behavior
kwargs passed in via low state via salt-api were not parsed correctly.
 
### New Behavior
yaml based kwargs are now handled correctly. see referenced issue for details.

### Tests written?
No, the existing tests should provide sufficient coverage for the behavior of localclient

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Review by gtmanfred — APPROVED** [review]
after looking at condition_input and parse_input (which calls condition_input) this looks fine to me.

File: salt/client/__init__.py [hunk 1]
```diff
@@ -322,7 +322,7 @@ def run_job(
             >>> local.run_job('*', 'test.sleep', [300])
             {'jid': '20131219215650131543', 'minions': ['jerry']}
         '''
-        arg = salt.utils.args.condition_input(arg, kwarg)
+        arg = salt.utils.args.parse_input(arg, kwarg)
```
Comment (gtmanfred) [inline salt/client/__init__.py]: parse_input takes `args` as one argument, this is passing kwarg to the `conditional` argument of parse_input

**Review by gtmanfred — CHANGES_REQUESTED** [review]

**Comment by mattp-** [comment]
Oops good catch. Let me try to sort out some unit tests to exercise this change. Can we put this on hold?

**Comment by rallytime** [comment]
@mattp- Sure thing. Let us know when you're ready.

**Review by isbm — APPROVED** [review]
:+1: with @gtmanfred though. Please fix this, otherwise OK.

File: salt/client/__init__.py (hunk 1 above)
Comment (mattp-) [reply salt/client/__init__.py]: i've updated it to a keyword arg

**Comment by mattp-** [comment]
@rallytime I've gone through and added unit tests to tests/unit/test_client.py, I believe this should be good now pending the jenkins runs. thanks!

**Comment by rallytime** [comment]
@mattp- Awesome! Thanks for writing those tests! It looks like you have some lint errors that need to be fixed up.

**Comment by mattp-** [comment]
yep my bad, fixed those

**Comment by mattp-** [comment]
spoke with @terminalmage / @cachedout on slack, I'm going to close this PR in favor of moving these changes into cherrypy/saltnado due to the sensitivity of localclient and how blunt yamlify'ing each arg is.

**Comment by mattp-** [comment]
reopening this for the time being until a decision on arg/kwarg until approach is made.

**Comment by mattp-** [comment]
note for myself, we need to account for   ```--no-parse=argname1,argname2,...
                        Comma-separated list of named CLI arguments (i.e.
                        argname=value) which should not be parsed as Python
                        data types```

**Comment by mattp-** [comment]
note 2: investigate if we can remove the redundant parse_args call in salt.utils.parser

**Comment by thatch45** [comment]
@mattp what is the status here?

**Comment by mattp-** [comment]
@thatch45 need to clean up some tests and remove the redundant kwarg calls in the cli parser before this is ready to go

**Comment by thatch45** [comment]
Fantastic! Thanks!

**Review by dwoz — APPROVED** [review]

**Pull request merged**